catch errors that happen outside of the normal error boundary

jivey opened the pull request:
For: https://github.com/openstax/unified/issues/1785

jivey committed:
catch errors that happen outside of the normal error boundary

jivey committed:
add test for onunhandledrejection

jivey committed:
fix lint

jivey committed:
fix alignment

jivey commented on src/app/messages/SimpleMessageProvider.tsx:
```diff
@@ -0,0 +1,27 @@
+  useEffect(() => {
+    const setUpIntl = async() => {
+      const languageWithCode = window!.navigator.language || 'en-US';
+      const language = new (Intl as any).Locale(languageWithCode).language;
```
The type for this is available in Typescript 4.5, so I figured it was desirable to wait to upgrade rather than pull in a package like `@formatjs/intl-locale`. But I can still pull that in, or regex it out instead.

TomWoodward replied:
if its on `window.Intl` you may just need to change `src/typings/global.d.ts` to expose it

we have some awkward typing going on around the browser globals to avoid them being defined when they shouldn't in ssr

jivey committed:
fix not-null assertion

staxly[bot] committed:
Merge branch 'master' into issue-1785

staxly[bot] committed:
Merge branch 'master' into issue-1785

staxly[bot] committed:
Merge branch 'master' into issue-1785

staxly[bot] committed:
Merge branch 'master' into issue-1785

staxly[bot] committed:
Merge branch 'master' into issue-1785

staxly[bot] committed:
Merge branch 'master' into issue-1785

jivey committed:
use assertWindow instead of window!

staxly[bot] committed:
Merge branch 'master' into issue-1785

staxly[bot] committed:
Merge branch 'master' into issue-1785

staxly[bot] committed:
Merge branch 'master' into issue-1785

staxly[bot] committed:
Merge branch 'master' into issue-1785

staxly[bot] committed:
Merge branch 'master' into issue-1785

staxly[bot] committed:
Merge branch 'master' into issue-1785

staxly[bot] committed:
Merge branch 'master' into issue-1785

staxly[bot] committed:
Merge branch 'master' into issue-1785

staxly[bot] committed:
Merge branch 'master' into issue-1785

staxly[bot] committed:
Merge branch 'master' into issue-1785

staxly[bot] committed:
Merge branch 'master' into issue-1785

staxly[bot] committed:
Merge branch 'master' into issue-1785

staxly[bot] committed:
Merge branch 'master' into issue-1785

staxly[bot] committed:
Merge branch 'master' into issue-1785

staxly[bot] committed:
Merge branch 'master' into issue-1785

staxly[bot] committed:
Merge branch 'master' into issue-1785

staxly[bot] committed:
Merge branch 'master' into issue-1785

staxly[bot] committed:
Merge branch 'master' into issue-1785

staxly[bot] committed:
Merge branch 'master' into issue-1785

staxly[bot] committed:
Merge branch 'master' into issue-1785

staxly[bot] committed:
Merge branch 'master' into issue-1785

staxly[bot] committed:
Merge branch 'master' into issue-1785

staxly[bot] committed:
Merge branch 'master' into issue-1785

staxly[bot] committed:
Merge branch 'master' into issue-1785

staxly[bot] committed:
Merge branch 'master' into issue-1785

staxly[bot] committed:
Merge branch 'master' into issue-1785

staxly[bot] committed:
Merge branch 'master' into issue-1785

staxly[bot] committed:
Merge branch 'master' into issue-1785

staxly[bot] committed:
Merge branch 'master' into issue-1785

staxly[bot] committed:
Merge branch 'master' into issue-1785

staxly[bot] committed:
Merge branch 'master' into issue-1785

staxly[bot] committed:
Merge branch 'master' into issue-1785

staxly[bot] committed:
Merge branch 'master' into issue-1785

staxly[bot] committed:
Merge branch 'master' into issue-1785

staxly[bot] committed:
Merge branch 'master' into issue-1785

staxly[bot] committed:
Merge branch 'master' into issue-1785

staxly[bot] committed:
Merge branch 'master' into issue-1785

staxly[bot] committed:
Merge branch 'master' into issue-1785

jivey committed:
send acceptStatus an error instance instead of a string

staxly[bot] committed:
Merge branch 'master' into issue-1785

staxly[bot] committed:
Merge branch 'master' into issue-1785

jivey committed:
fix lint

jivey committed:
add Locale typing to globals

jivey committed:
use assertWindow

staxly[bot] committed:
Merge branch 'master' into issue-1785

staxly[bot] committed:
Merge branch 'master' into issue-1785

staxly[bot] committed:
Merge branch 'master' into issue-1785

staxly[bot] committed:
Merge branch 'master' into issue-1785

jivey committed:
update EN error page text

bethshook commented on src/app/messages/en/modalErrorText.ts:
```diff
@@ -1,4 +1,4 @@
 export const bodyWithLink = `
-  We're not quite sure what went wrong. Refresh the page. If this doesn't solve
+  We're not quite sure what went wrong. Restart your browser. If this doesn't solve
```
For Spanish, replace "Actualice la página." with "Reinicia el navegador."

jivey replied:
Thanks!

jivey committed:
update translation

staxly[bot] committed:
Merge branch 'master' into issue-1785

jivey committed:
add snapshot for promise error render

bethshook commented on src/app/errors/components/ErrorBoundary.spec.tsx:
```diff
@@ -41,4 +54,17 @@ describe('ErrorBoundary', () => {
   });
+
+  it('captures unhandled rejected promises', () => {
+    const tree = renderer.create(<TestContainer>
```
maybe this doesn't matter here but <ErrorBoundary> would not normally be wrapped in this container since TestContainer includes Services and MessageProvider

jivey replied:
I think the ErrorBoundary needs the MessageProvider, I'm not sure if that's always been the case, but it complains if I remove it

bethshook replied:
Yeah, it's probably fine for the purposes of this test - it just technically should be using your new SimpleMessageProvider since ErrorBoundary falls outside of the normal MessageProvider (which gets its language from whatever book is loaded)

staxly[bot] committed:
Merge branch 'master' into issue-1785

jivey committed:
debug sentry call

staxly[bot] committed:
Merge branch 'master' into issue-1785

jivey committed:
remove debug sentry

staxly[bot] committed:
Merge branch 'master' into issue-1785

jivey committed:
catch and ignore errors from forked promise to fix double throw

Before this change, if a collected promise threw an error, it
would throw twice because of .then/.finally forking the promise.
But we don't actually care about the error here (the original
will still bubble and be handled) so swallow the error so it is
ignored.

jivey committed:
update error text

jivey committed:
skip null intl conditional for prerender scenario

jivey committed:
fix lint

jivey committed:
check for window def before adding listener

jivey committed:
fix lint

jivey committed:
fix lint

staxly[bot] committed:
Merge branch 'master' into issue-1785

jivey committed:
add a test case for promise collector bug fix

Without the .catch that was added to PromiseCollector.add, the
.finally chain fork will cause rejections to occur twice and
bubble up to onunhandledrejection

jivey committed:
fix lint

jivey committed:
remove incorrect buggy promise error test

staxly[bot] committed:
Merge branch 'master' into issue-1785

staxly[bot] committed:
Merge branch 'master' into issue-1785

jivey committed:
use a fake PromiseRejectionEvent to test boundary

jivey committed:
Merge branch 'master' into issue-1785

staxly[bot] committed:
Merge branch 'master' into issue-1785

jivey committed:
fix lint

jivey committed:
add test case for en default

jivey committed:
Merge branch 'master' into issue-1785

staxly[bot] committed:
Merge branch 'master' into issue-1785

staxly[bot] committed:
Merge branch 'master' into issue-1785

staxly[bot] committed:
Merge branch 'master' into issue-1785

TomWoodward commented on src/app/errors/components/ErrorBoundary.tsx:
```diff
@@ -56,6 +60,18 @@ class ErrorBoundary extends React.Component<Props, State> {
 
+  public componentDidMount() {
+    if (windowDefined()) {
+      window?.addEventListener('unhandledrejection', this.handleRejection);
```
if you just do the `typeof` check here in the if statement, you won't need the `?`, i don't love the creation of the extra code path

jivey commented on src/helpers/PromiseCollector.spec.ts:
```diff
@@ -58,4 +58,14 @@ describe('PromiseCollector', () => {
   });
+
+  it('swallows errors when removing promises with .finally', async() => {
+    // The "fail" case of this test is with a process exit code
```
I had wondered the same, and found out something interesting: jest creates its own process so it can attach handlers, which makes it difficult/awkward to work around. I found some interesting looking suggestions on the jest issue, like https://github.com/facebook/jest/issues/5620#issuecomment-761879230 and https://github.com/facebook/jest/issues/5620#issuecomment-998716759 but wasn't sure we'd want to complicate the test suite like that. I could explore it more though and try to find something that feels like a good fit.

jivey committed:
inline typeof check

jivey committed:
add clarifying comment for why we re-throw ArchiveBookMissingError

jivey committed:
move locale type into window

jivey committed:
improve comment wording

staxly[bot] committed:
Merge branch 'master' into issue-1785

staxly[bot] committed:
Merge branch 'master' into issue-1785

staxly[bot] committed:
Merge branch 'master' into issue-1785

staxly[bot] committed:
Merge branch 'master' into issue-1785

staxly[bot] committed:
Merge branch 'master' into issue-1785

TomWoodward reviewed (approved):
this looks good, but needs conflicts resolved

Malar-Natarajan commented:
@jivey looks like this is having merge conflicts

jivey committed:
Merge branch 'master' into issue-1785
